src/packages/client/src/runtime/utils/validatePrismaClientOptions.ts

```diff
@@ -0,0 +1,169 @@
+  datasources: (options: any, datasourceNames: string[]) => {
+    if (!options) {
+      return
+    }
+    if (
+      typeof options !== 'object' || Array.isArray(options)
+    ) {
+      throw new PrismaClientValidationError(`Invalid value ${JSON.stringify(options)} for "datasources" provided to PrismaClient constructor`)
+    }
+
+    for (const [key, value] of Object.entries(options)) {
+      if (!datasourceNames.includes(key)) {
+        const didYouMean = getDidYouMean(key, datasourceNames) || `Available datasources: ${datasourceNames.join(', ')}`
+        throw new PrismaClientValidationError(`Unknown datasource ${key} provided to PrismaClient constructor.${didYouMean}`)
+      }
+      if (typeof value !== 'object' || Array.isArray(value)) {
+        throw new PrismaClientValidationError(
+          `Invalid value ${JSON.stringify(options)} for datasource "${key}" provided to PrismaClient constructor.
+It should have this form: { url: "CONNECTION_STRING" }`
+        )
+      }
+      if (value && typeof value === 'object') {
+        for (const [key1, value1] of Object.entries(value)) {
+          if (key1 !== 'url') {
+            throw new PrismaClientValidationError(
+              `Invalid value ${JSON.stringify(options)} for datasource "${key}" provided to PrismaClient constructor.
+It should have this form: { url: "CONNECTION_STRING" }`
+            )
+          }
+          if (typeof value1 !== 'string') {
+            throw new PrismaClientValidationError(
+              `Invalid value ${JSON.stringify(value1)} for datasource "${key}" provided to PrismaClient constructor.
+It should have this form: { url: "CONNECTION_STRING" }`
+            )
+          }
+        }
+      }
+    }
+  },
+  errorFormat: (options: any) => {
+    if (!options) {
+      return
+    }
+    if (typeof options !== 'string') {
+      throw new PrismaClientValidationError(`Invalid value ${JSON.stringify(options)} for "errorFormat" provided to PrismaClient constructor.`)
+    }
+    if (!errorFormats.includes(options as ErrorFormat)) {
+      const didYouMean = getDidYouMean(options, errorFormats)
+      throw new PrismaClientValidationError(`Invalid errorFormat ${options} provided to PrismaClient constructor.${didYouMean}`)
+    }
+  },
+  log: (options: any) => {
+    if (!options) {
+      return
+    }
+    if (!Array.isArray(options)) {
+      throw new PrismaClientValidationError(`Invalid value ${JSON.stringify(options)} for "log" provided to PrismaClient constructor.`)
+    }
+
+    function validateLogLevel(level: any) {
+      if (typeof level === 'string') {
+        if (!logLevels.includes(level as LogLevel)) {
+          const didYouMean = getDidYouMean(level, logLevels)
+          throw new PrismaClientValidationError(`Invalid log level "${level}" provided to PrismaClient constructor.${didYouMean}`)
+        }
+      }
+    }
+
+    for (const option of options) {
+      validateLogLevel(option)
+
+      const validators = {
```

Comment:
Could maybe be renamed to `logValidators` as there is already a top level validators const?